src/java/htsjdk/samtools/util/BlockCompressedInputStream.java

```diff
@@ -507,6 +542,117 @@ private static boolean preambleEqual(final byte[] preamble, final byte[] buf, fi
         }
         return true;
     }
+    
+    private static class DecompressedBlock {
+        /**
+         * Decompressed block
+         */
+        private final byte[] mBlock;
+        /**
+         * Compressed size of block (the uncompressed size can be found using mBlock.length)
+         */
+        private final int mBlockCompressedSize;
+        /**
+         * Stream offset of start of block
+         */
+        private final long mBlockAddress;
+        /**
+         * Exception thrown (if any) when attempting to decompress block
+         */
+        private final Exception mException;
+        public DecompressedBlock(long blockAddress, byte[] block, int compressedSize) {
+            mBlock = block;
+            mBlockAddress = blockAddress;
+            mBlockCompressedSize = compressedSize;
+            mException = null;
+        }
+        public DecompressedBlock(long blockAddress, int compressedSize, Exception exception) {
+            mBlock = new byte[0];
+            mBlockAddress = blockAddress;
+            mBlockCompressedSize = compressedSize;
+            mException = exception;
+        }
+    }
+    
+    private void startDecompressionThread() {
+        if (mDecompressionThread == null) {
+            mDecompressionRunnable = new DecompressionRunnable();
+            mDecompressionThread = new Thread(mDecompressionRunnable, "DecompressBGZF");
+            mDecompressionThread.setDaemon(true);
+            mDecompressionThread.start();
```

Comment:
Rather than expanding, I'd rather shrink the PR to split it to changes in BlockCompressedInputStream.java+related and separate PR for AsyncBufferedIterator. Having small PRs would greatly simplify our life and enable faster merging.

Using a thread pool for async writing would be fabulous (see also #559 ) but definitely a separate PR.